# Spotify Music Recommendation System with Mood Detection Using The Million Song Dataset

Author: James Meredith

Instructor: David Elliot

Institution: Flatiron School

Active Project Dates: Jun 267th, 2023 - July 14th, 2023

## Overview

Recommendation systems use machine learning to solve the long-tail problem with digital content. There is a lot of content out on the internet so it would take a person a long time to sort through it to find something they like. Recommendation systems help wth that buy suggesting content to people. [Netflix estimates](https://towardsdatascience.com/deep-dive-into-netflixs-recommender-system-341806ae3b48) it gets around 80% of total watch time thanks to its recommendation system.

The data is fairly clean already so most of the preperation involved manipulating the datasets into formats that can be used by recommendation systems, some recommendation systems use the format provided but ALS by hand and cosign similarity required some modifications.

Our modeling approach was:

- Start with a content-based system using cosign similarity, a commonly used distance measurement for recommendation systems
- Move to collaborative filtering using surprise, the popular library for implementing predictive recommendation systems
- Introduce ALS by hand to cement our understanding of the approach
- Implement ALS in Spark using pyspark separately, spark is commonly used to implement ALS at scale

We evaluated these models using RMSE as the measurement. Netflix's famous competition that offered $1M to a winner that could improve their content recommendation algorythm by 10% RMSE which was finally won by BellKor’s Pragmatic Chaos in 2009 with an RMSE of 0.8567. We decided that was our goal. All models were trained on train data, cross validated, and then evaluated on test data.

Note: ratings were on a scale of 1-5 so a RMSE of 1 means the average prediction was off by a whole score

Our best performing model was SVD optimized using gridsearch using surprise. Our recommendation however is to use ALS via spark to start however because of the efficiency of ALS and Spark in handling large datasets.

## Business Problem

## The Data

The dataset used for this project was the Million Song Database. It contains metadata for one million songs. The metadata includes information such as artist, title, year, and genre. The dataset also contains audio features for each song. The audio features are extracted from the audio files using the Echo Nest analyzer. The audio features include information such as tempo, loudness, and key. The dataset also contains lyrics for each song. The lyrics are extracted from the audio files using the Musixmatch analyzer. The lyrics are stored as a bag of words. The dataset is stored in HDF5 format. The dataset is 280 GB uncompressed. The dataset is available for download at http://millionsongdataset.com/. The dataset is also available on AWS S3 at s3://millionsongdataset/. 

The Million Song Dataset is a freely-available collection of audio features and metadata for a million contemporary popular music tracks. The Million Song Dataset started as a collaborative project between The Echo Nest and LabROSA. It was supported in part by the NSF. 

Its purposes are:

- To encourage research on algorithms that scale to commercial sizes
- To provide a reference dataset for evaluating research
- As a shortcut alternative to creating a large dataset with APIs (e.g. The Echo Nest's)
- To help new researchers get started in the MIR field

The core of the dataset is the feature analysis and metadata for one million songs, provided by The Echo Nest. 

Thierry Bertin-Mahieux, Daniel P.W. Ellis, Brian Whitman, and Paul Lamere. 
The Million Song Dataset. In Proceedings of the 12th International Society
for Music Information Retrieval Conference (ISMIR 2011), 2011.

### Exploring The Dataset

In [149]:
# Imports Required for Accessing the Data
import os
import sys
import time
import glob
import datetime
import sqlite3
import numpy as np
import pandas as pd
import tables as tb
import h5py
import hdf5_getters as GETTERS


# Standard imports
from random import gauss, uniform as uni, seed
import math as math
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Cosign Similarity
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.preprocessing import StandardScaler

# Surprise imports for collaborative filtering
from surprise import Dataset, Reader
from surprise.prediction_algorithms import *
from surprise.model_selection import cross_validate, train_test_split, GridSearchCV
from surprise.accuracy import rmse, mae

# ALS
from sklearn.linear_model import LinearRegression

The structure of the dataset is quite complex, with many nested folders with files within them. The first step is to explore the structure of the dataset. Here we use the os module to explore the structure of the dataset.

In [40]:
# Establishes path to the Million Song Dataset subset
msd_subset_path='./MillionSongSubset'
msd_subset_data_path=os.path.join(msd_subset_path,'data')
msd_subset_addf_path=os.path.join(msd_subset_path,'AdditionalFiles')

# Establishes path to the Million Song Dataset code
msd_code_path='./MSongsDB'
assert os.path .isdir(msd_code_path),'wrong path' # sanity check
# we add some paths to python so we can import MSD code
# Ubuntu: you can change the environment variable PYTHONPATH
# in your .bashrc file so you do not have to type these lines
sys.path.append( os.path.join(msd_code_path,'PythonSrc') )

# Creates a function to display the folder structure of the dataset
def display_folder_structure(path, indent='', max_depth=float('inf'), depth=0):
    if depth > max_depth:
        return

    files = os.listdir(path)
    for file in files:
        current_path = os.path.join(path, file)
        if os.path.isdir(current_path):
            print(f"{indent}|- {file}/")
            display_folder_structure(current_path, indent + '  ', max_depth, depth + 1)
        else:
            print(f"{indent}|- {file}")


# Uses the function to display the folder structure of the dataset to the specified depth
max_display_depth = 0  # Specify the maximum depth to display
display_folder_structure(msd_subset_path, max_depth=max_display_depth)

|- A/
|- B/


So the structure of the dataset is as follows:

- The dataset is stored in the folder `MillionSongSubset`
- The individual dataset files are stored in HDF5 format
- The dataset is stored in a nested folder structure
- The dataset contains 10,000 folders, each containing 100 folders, each containing 100 files

The next step is to explore the contents of the dataset. Here we use the h5py module to explore the contents of the dataset.


In [41]:
# Let's open one file and explore it's structure and contents
filename = '.\MillionSongSubset\A\A\A\TRAAAAW128F429D538.h5'
h5file = tb.open_file(filename, mode='r')
h5file 

File(filename=.\MillionSongSubset\A\A\A\TRAAAAW128F429D538.h5, title='H5 Song File', mode='r', root_uep='/', filters=Filters(complevel=1, complib='zlib', shuffle=True, bitshuffle=False, fletcher32=False, least_significant_digit=None))
/ (RootGroup) 'H5 Song File'
/analysis (Group) 'Echo Nest analysis of the song'
/analysis/bars_confidence (EArray(83,)shuffle, zlib(1)) 'array of confidence of bars'
  atom := Float64Atom(shape=(), dflt=0.0)
  maindim := 0
  flavor := 'numpy'
  byteorder := 'little'
  chunkshape := (1024,)
/analysis/bars_start (EArray(83,)shuffle, zlib(1)) 'array of start times of bars'
  atom := Float64Atom(shape=(), dflt=0.0)
  maindim := 0
  flavor := 'numpy'
  byteorder := 'little'
  chunkshape := (1024,)
/analysis/beats_confidence (EArray(344,)shuffle, zlib(1)) 'array of confidence of sections'
  atom := Float64Atom(shape=(), dflt=0.0)
  maindim := 0
  flavor := 'numpy'
  byteorder := 'little'
  chunkshape := (1024,)
/analysis/beats_start (EArray(344,)shuffle, zlib(1

In [42]:
h5file.root

/ (RootGroup) 'H5 Song File'
  children := ['analysis' (Group), 'metadata' (Group), 'musicbrainz' (Group)]

In [43]:
h5file.root.analysis


/analysis (Group) 'Echo Nest analysis of the song'
  children := ['bars_confidence' (EArray), 'bars_start' (EArray), 'beats_confidence' (EArray), 'beats_start' (EArray), 'sections_confidence' (EArray), 'sections_start' (EArray), 'segments_confidence' (EArray), 'segments_loudness_max' (EArray), 'segments_loudness_max_time' (EArray), 'segments_loudness_start' (EArray), 'segments_pitches' (EArray), 'segments_start' (EArray), 'segments_timbre' (EArray), 'songs' (Table), 'tatums_confidence' (EArray), 'tatums_start' (EArray)]

In [44]:
h5file.root.metadata

/metadata (Group) 'metadata about the song'
  children := ['artist_terms' (EArray), 'artist_terms_freq' (EArray), 'artist_terms_weight' (EArray), 'similar_artists' (EArray), 'songs' (Table)]

In [45]:
h5file.root.musicbrainz

/musicbrainz (Group) 'data about the song coming from MusicBrainz'
  children := ['artist_mbtags' (EArray), 'artist_mbtags_count' (EArray), 'songs' (Table)]

So the structure of each .h5 file is as follows:
- The file contains 3 groups: analysis, metadata, and musicbrainz
- The analysis group contains:
    bars_confidence, bars_start, beats_confidence, beats_start, sections_confidence, sections_start, segments_confidence, segments_loudness_max, segments_loudness_max_time, segments_loudness_start, segments_pitches, segments_start, segments_timbre, songs, tatums_confidence, and tatums_start
- The metadata group contains:
    artist_terms, artist_terms_freq, artist_terms_weight, similar_artists, songs, and tags
- The musicbrainz group contains:
    artist_mbtags, artist_mbtags_count, songs, and tags


## Data Exploration

## Data Preperation & Cleaning

Because the structure of the dataset is catered towards big data applications, it is not ideal for machine learning applications. The first step is to convert the dataset into a more machine learning friendly format. Here we use the pandas module to convert the dataset into a pandas dataframe.

In [46]:
df = pd.DataFrame(h5file)
df

0    \
0   /analysis (Group) 'Echo Nest analysis of the s...   
1   /analysis/bars_confidence (EArray(83,)shuffle,...   
2   /metadata/artist_terms (EArray(37,)shuffle, zl...   
3   /musicbrainz/artist_mbtags (EArray(0,)shuffle,...   
4                                               0.643   
5                                             0.58521   
6                                               0.834   
7                                             0.58521   
8                                                 1.0   
9                                                 0.0   
10                                                0.0   
11                                              -60.0   
12                                                0.0   
13                                              -60.0   
14  [0.946, 0.684, 0.679, 0.941, 0.744, 0.633, 0.7...   
15                                                0.0   
16  [0.0, 171.13, 9.469, -28.48, 57.491, -50.067, ...   
17      /analysis/songs.row (Row), pointing to row #0   
18                                              0.779   
19                                            0.28519   
20                                         b'hip hop'   
21                                                1.0   
22                                                1.0   
23                              b'ARV4KO21187FB38008'   
24      /metadata/songs.row (Row), pointing to row #0   
25                                               None   
26                                               None   
27   /musicbrainz/songs.row (Row), pointing to row #0   

                                                  1    \
0         /metadata (Group) 'metadata about the song'   
1   /analysis/bars_start (EArray(83,)shuffle, zlib...   
2   /metadata/artist_terms_freq (EArray(37,)shuffl...   
3   /musicbrainz/artist_mbtags_count (EArray(0,)sh...   
4                                               0.746   
5                                             2.94247   
6                                               0.851   
7                                             1.19196   
8                                                 1.0   
9                                             7.74554   
10                                                1.0   
11                                            -31.646   
12                                            0.10929   
13                                              -60.0   
14  [0.01, 0.054, 0.015, 0.021, 0.067, 0.17, 1.0, ...   
15                                            0.24671   
16  [19.991, -143.504, -118.249, -142.909, -18.528...   
17                                               None   
18                                              0.734   
19                                            0.58521   
20                                 b'underground rap'   
21                                           0.776136   
22                                           0.897936   
23                              b'ARWHM281187FB3D381'   
24                                               None   
25                                               None   
26                                               None   
27                                               None   

                                                  2    \
0   /musicbrainz (Group) 'data about the song comi...   
1   /analysis/beats_confidence (EArray(344,)shuffl...   
2   /metadata/artist_terms_weight (EArray(37,)shuf...   
3   [/musicbrainz/songs.row (Row), pointing to row...   
4                                               0.722   
5                                             5.14371   
6                                                0.65   
7                                             1.78893   
8                                               0.218   
9                                            36.44331   
10                                              0.483   
11                                            -34.565   
1

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Columns: 971 entries, 0 to 970
dtypes: object(971)
memory usage: 212.5+ KB


Attempting to import an individual data file wholesale into a pandas dataframe results in a dataframe with 971 columns and 27 rows - for just one song! Multiply that by 10,000 for the number of songs in the dataset and you have a dataframe with 9,710,000 columns and 270,000 rows - for just the Million Song Subset! This is not ideal for loading into memory on a local machine. Instead we will only load the columns we need into the dataframe moving forward. We'll start by defining a function that can iterate through the dataset and load the columns we need into a dataframe, and then prepare the dataframe for machine learning applications.

In [95]:
# we define this very useful function to iterate through all the files in the dataset
def apply_to_all_files(basedir,func=lambda x: x,ext='.h5'):
    """
    From a base directory, go through all subdirectories,
    find all files with the given extension, apply the
    given function 'func' to all of them.
    If no 'func' is passed, we do nothing except counting.
    INPUT
       basedir  - base directory of the dataset
       func     - function to apply to all filenames
       ext      - extension, .h5 by default
    RETURN
       number of files
    """
    cnt = 0
    # iterate over all files in all subdirectories
    for root, dirs, files in os.walk(basedir):
        files = glob.glob(os.path.join(root,'*'+ext))
        # count files
        cnt += len(files)
        # apply function to all files
        for f in files :
            func(f)       
    return cnt

In [96]:
# we can now easily count the number of files in the dataset
print('number of song files:',apply_to_all_files(msd_subset_path))

number of song files: 10000


In [139]:
df = pd.DataFrame()

In [141]:
# Creates a function to open a file and return the data
def get_data(filename):
    # Opens the file
    get_data_file = GETTERS.open_h5_file_read(filename)
    # Append the song ID to the dataframe as the index
    df.loc[filename, 'song_id'] = GETTERS.get_song_id(get_data_file).decode('utf-8')
    # Append the artist ID to the dataframe
    df.loc[filename, 'artist_id'] = GETTERS.get_artist_id(get_data_file).decode('utf-8')
    # Append the release to the dataframe
    df.loc[filename, 'release'] = GETTERS.get_release(get_data_file).decode('utf-8')
    # Appends the song year to the dataframe
    df.loc[filename, 'year'] = GETTERS.get_year(get_data_file)
    # Appends the song mode to the dataframe
    df.loc[filename, 'mode'] = GETTERS.get_mode(get_data_file)
    # Appends the song time signature to the dataframe
    df.loc[filename, 'time_signature'] = GETTERS.get_time_signature(get_data_file)
    # Appends the song tempo to the dataframe
    df.loc[filename, 'tempo'] = GETTERS.get_tempo(get_data_file)
    # Appends the song loudness to the dataframeq
    df.loc[filename, 'loudness'] = GETTERS.get_loudness(get_data_file)
    # # Appends the song energy to the dataframe - omitted due to lack of data
    # df.loc[filename, 'energy'] = GETTERS.get_energy(get_data_file)
    # # Appends the song danceability to the dataframe - omitted due to lack of data
    # df.loc[filename, 'danceability'] = GETTERS.get_danceability(get_data_file) 
    # Appends the popularity to the dataframe
    # df.loc[filename, 'song_hotttnesss'] = GETTERS.get_song_hotttnesss(get_data_file)
    #closes the file
    get_data_file.close()

In [142]:
# Applies the function to all files in the dataset
apply_to_all_files(msd_subset_path, func=get_data)

10000

In [143]:
df.head(20)

song_id  \
./MillionSongSubset\A\A\A\TRAAAAW128F429D538.h5  SOMZWCG12A8C13C480   
./MillionSongSubset\A\A\A\TRAAABD128F429CF47.h5  SOCIWDW12A8C13D406   
./MillionSongSubset\A\A\A\TRAAADZ128F9348C2E.h5  SOXVLOJ12AB0189215   
./MillionSongSubset\A\A\A\TRAAAEF128F4273421.h5  SONHOTT12A8C13493C   
./MillionSongSubset\A\A\A\TRAAAFD128F92F423A.h5  SOFSOCN12A8C143F5D   
./MillionSongSubset\A\A\A\TRAAAMO128F1481E7F.h5  SOYMRWW12A6D4FAB14   
./MillionSongSubset\A\A\A\TRAAAMQ128F1460CD3.h5  SOMJBYD12A6D4F8557   
./MillionSongSubset\A\A\A\TRAAAPK128E0786D96.h5  SOHKNRJ12A6701D1F8   
./MillionSongSubset\A\A\A\TRAAARJ128F9320760.h5  SOIAZJW12AB01853F1   
./MillionSongSubset\A\A\A\TRAAAVG12903CFA543.h5  SOUDSGM12AC9618304   
./MillionSongSubset\A\A\A\TRAAAVO128F93133D4.h5  SOQHXMF12AB0182363   
./MillionSongSubset\A\A\B\TRAABCL128F4286650.h5  SOKEJEJ12A8C13E0D0   
./MillionSongSubset\A\A\B\TRAABDL12903CAABBA.h5  SOMUYGI12AB0188633   
./MillionSongSubset\A\A\B\TRAABJL12903CDCF1A.h5  SORAMLE12AB017C8B0   
./MillionSongSubset\A\A\B\TRAABJV128F1460C49.h5  SOBONFF12A6D4F84D8   
./MillionSongSubset\A\A\B\TRAABLR128F423B7E3.h5  SOHUOAP12A8AE488E9   
./MillionSongSubset\A\A\B\TRAABNV128F425CEE1.h5  SOUQQEA12A8C134B1B   
./MillionSongSubset\A\A\B\TRAABRB128F9306DD5.h5  SOILPQQ12AB017E82A   
./MillionSongSubset\A\A\B\TRAABVM128F92CA9DC.h5  SOXLBJT12A8C140925   
./MillionSongSubset\A\A\B\TRAABXG128F9318EBD.h5  SOZVMJI12AB01808AF   

                                                          artist_id  \
./MillionSongSubset\A\A\A\TRAAAAW128F429D538.h5  ARD7TVE1187B99BFB1   
./MillionSongSubset\A\A\A\TRAAABD128F429CF47.h5  ARMJAGH1187FB546F3   
./MillionSongSubset\A\A\A\TRAAADZ128F9348C2E.h5  ARKRRTF1187B9984DA   
./MillionSongSubset\A\A\A\TRAAAEF128F4273421.h5  AR7G5I41187FB4CE6C   
./MillionSongSubset\A\A\A\TRAAAFD128F92F423A.h5  ARXR32B1187FB57099   
./MillionSongSubset\A\A\A\TRAAAMO128F1481E7F.h5  ARKFYS91187B98E58F   
./MillionSongSubset\A\A\A\TRAAAMQ128F1460CD3.h5  ARD0S291187B9B7BF5   
./MillionSongSubset\A\A\A\TRAAAPK128E0786D96.h5  AR10USD1187B99F3F1   
./MillionSongSubset\A\A\A\TRAAARJ128F9320760.h5  AR8ZCNI1187B9A069B   
./MillionSongSubset\A\A\A\TRAAAVG12903CFA543.h5  ARNTLGG11E2835DDB9   
./MillionSongSubset\A\A\A\TRAAAVO128F93133D4.h5  ARGSJW91187B9B1D6B   
./MillionSongSubset\A\A\B\TRAABCL128F4286650.h5  ARC43071187B990240   
./MillionSongSubset\A\A\B\TRAABDL12903CAABBA.h5  ARL7K851187B99ACD2   
./MillionSongSubset\A\A\B\TRAABJL12903CDCF1A.h5  ARHHO3O1187B989413   
./MillionSongSubset\A\A\B\TRAABJV128F1460C49.h5  ARIK43K1187B9AE54C   
./MillionSongSubset\A\A\B\TRAABLR128F423B7E3.h5  ARD842G1187B997376   
./MillionSongSubset\A\A\B\TRAABNV128F425CEE1.h5  ARIG6O41187B988BDD   
./MillionSongSubset\A\A\B\TRAABRB128F9306DD5.h5  AR1ZHYZ1187FB3C717   
./MillionSongSubset\A\A\B\TRAABVM128F92CA9DC.h5  ARYKCQI1187FB3B18F   
./MillionSongSubset\A\A\B\TRAABXG128F9318EBD.h5  ARNPAGP1241B9C7FD4   

                                                                             release  \
./MillionSongSubset\A\A\A\TRAAAAW128F429D538.h5                          Fear Itself   
./MillionSongSubset\A\A\A\TRAAABD128F429CF47.h5                           Dimensions   
./MillionSongSubset\A\A\A\TRAAADZ128F9348C2E.h5  Las Numero 1 De La Sonora Santanera   
./MillionSongSubset\A\A\A\TRAAAEF128F4273421.h5                        Friend Or Foe   
./MillionSongSubset\A\A\A\TRAAAFD128F92F423A.h5                        Muertos Vivos   
./MillionSongSubset\A\A\A\TRAAAMO128F1481E7F.h5                         Ordinary Day   
./MillionSongSubset\A\A\A\TRAAAMQ128F1460CD3.h5                    Da Ghetto Psychic   
./MillionSongSubset\A\A\A\TRAAAPK128E0786D96.h5                         Gin & Phonic   
./MillionSongSubset\A\A\A\TRAAARJ128F9320760.h5                           Pink World   
./MillionSongSubset\A\A\A\TRAAAVG12903CFA543.h5                    Superinstrumental   
./MillionSongSubset\A\A\A\TRAAAVO128F93133D4.h5                           I Need You   
./MillionSongSubset\

In [144]:
# Sets the index to the song ID
df_test = df.set_index('song_id')

## Modeling

### Content-Based Model using Cosine Similarity

In [ ]:
features = ['mode', 'time_signature', 'tempo', 'loudness']

In [ ]:
scaler = StandardScaler()
df_test_scaled = scaler.fit_transform(df_test[features])

In [ ]:
cosine_sim = cosine_similarity(df_test_scaled)

In [ ]:
similarity_matrix = pd.DataFrame(cosine_sim, index=df['song_id'], columns=df['song_id'])


In [ ]:
similarity_matrix.head(20)

song_id             SOMZWCG12A8C13C480  SOCIWDW12A8C13D406  \
song_id                                                      
SOMZWCG12A8C13C480            1.000000            0.872983   
SOCIWDW12A8C13D406            0.872983            1.000000   
SOXVLOJ12AB0189215           -0.290332           -0.478978   
SONHOTT12A8C13493C           -0.579917           -0.680180   
SOFSOCN12A8C143F5D           -0.499656           -0.364192   
SOYMRWW12A6D4FAB14           -0.930551           -0.690332   
SOMJBYD12A6D4F8557           -0.283379           -0.475992   
SOHKNRJ12A6701D1F8            0.884208            0.801997   
SOIAZJW12AB01853F1            0.037200           -0.419751   
SOUDSGM12AC9618304            0.821943            0.932622   
SOQHXMF12AB0182363            0.203017            0.378203   
SOKEJEJ12A8C13E0D0           -0.499905           -0.560551   
SOMUYGI12AB0188633           -0.616390           -0.495747   
SORAMLE12AB017C8B0            0.242438            0.395002   
SOBONFF12A6D4F84D8           -0.793414           -0.801640   
SOHUOAP12A8AE488E9           -0.575852           -0.649532   
SOUQQEA12A8C134B1B           -0.198986           -0.285651   
SOILPQQ12AB017E82A           -0.181295           -0.467586   
SOXLBJT12A8C140925           -0.656896           -0.360355   
SOZVMJI12AB01808AF           -0.690882           -0.481267   

song_id             SOXVLOJ12AB0189215  SONHOTT12A8C13493C  \
song_id                                                      
SOMZWCG12A8C13C480           -0.290332           -0.579917   
SOCIWDW12A8C13D406           -0.478978           -0.680180   
SOXVLOJ12AB0189215            1.000000           -0.078905   
SONHOTT12A8C13493C           -0.078905            1.000000   
SOFSOCN12A8C143F5D           -0.069991            0.776698   
SOYMRWW12A6D4FAB14            0.378492            0.317649   
SOMJBYD12A6D4F8557            0.818307           -0.278453   
SOHKNRJ12A6701D1F8            0.071113           -0.843271   
SOIAZJW12AB01853F1            0.108605            0.462460   
SOUDSGM12AC9618304           -0.374759           -0.483023   
SOQHXMF12AB0182363            0.458925           -0.828339   
SOKEJEJ12A8C13E0D0           -0.037049            0.961857   
SOMUYGI12AB0188633           -0.116711            0.867705   
SORAMLE12AB017C8B0            0.500981           -0.763063   
SOBONFF12A6D4F84D8            0.680840            0.545356   
SOHUOAP12A8AE488E9           -0.075027            0.995597   
SOUQQEA12A8C134B1B           -0.555962            0.396466   
SOILPQQ12AB017E82A            0.102828            0.850297   
SOXLBJT12A8C140925           -0.187997            0.642142   
SOZVMJI12AB01808AF           -0.174860            0.791626   

song_id             SOFSOCN12A8C143F5D  SOYMRWW12A6D4FAB14  \
song_id                                                      
SOMZWCG12A8C13C480           -0.499656           -0.930551   
SOCIWDW12A8C13D406           -0.364192           -0.690332   
SOXVLOJ12AB0189215           -0.069991            0.378492   
SONHOTT12A8C13493C            0.776698            0.317649   
SOFSOCN12A8C143F5D            1.000000            0.453094   
SOYMRWW12A6D4FAB14            0.453094            1.000000   
SOMJBYD12A6D4F8557           -0.505654            0.313475   
SOHKNRJ12A6701D1F8           -0.659999           -0.693216   
SOIAZJW12AB01853F1           -0.134141           -0.365149   
SOUDSGM12AC9618304           -0.066293           -0.633734   
SOQHXMF12AB0182363           -0.416180            0.140855   
SOKEJEJ12A8C13E0D0            0.898502            0.304083   
SOMUYGI12AB0188633            0.976352            0.512086   
SORAMLE12AB017C8B0           -0.299800            0.106467   
SOBONFF12A6D4F84D8            0.613140            0.792289   
SOHUOAP12A8AE488E9            0.831572            0.337425   
SOUQQEA12A8C134B1B           -0.127822           -0.101010   
SOILPQQ12AB017E82A            0.627055           -0.069117   
SOXLBJT12A8C140925            

### Collaborative Filtering

### Aleternating Least-Squares (ALS)

## Conclusion:

- Content based systems are simple to implement but hard to measure error (no RMSE)
- Collaborative filtering has issues with cold start (movies with no ratings)
- Collaborative filtering using surprise is incredibly effective and produces great models
- ALS is incredibly efficient an simple

There are many options to create recommendation systems. 
In this notebook we explored content based systems using cosign similarity, 
various collaborative filtering models via surprise,
and dove deep into the concepts behind ALS to explain how it works.

We were able to beat BellKor’s Pragmatic Chaos's RMSE of 0.8567 with an RMSE of 0.8559 but this was likely because of our smaller dataset.

The best approach to recommendation systems is likely a hybrid approach which we will not explore in this notebook.